### PR TITLE
[8.0] Refactor lettermgmt

### PR DESCRIPTION
**Pull request by ivantodorovich**
As discussed in https://github.com/OCA/crm/issues/83

Feel free to comment and make suggestions as I work.

TODO:
- [x] ~~Migrate field class to class_id~~ _migrated to category_ids._
- [x] ~~Add seccurity group to show/hide reassignments~~
- [x] ~~Add seccurity group to show/hide partner / child~~
- [x] ~~Increment sequence only after the record is saved, not before.~~
- [x] ~~Simplify and improve state workflow.~~
- [ ] ~~Support current "advanced" workflow in a separate module.~~ _wont support it in this pr_
- [x] ~~Add default filter 'Pending' and search fields to list views.~~
- [x] ~~Add tests~~

### CODE REVIEW

**Comment by hbrunn** [comment]
another great thing to add would be tests...

**Comment by ivantodorovich** [comment]
@hbrunn 
Please take a look when you have the time.

**Functional changes:**
- Field `class` is now a Many2many_tags called `category_ids`
- Field `date` is now field.Date (instead of Datetime).
- Tree View improvements
  - Added colours
  - Added a default filter 'Pending' to show letters that are not received yet (either by us or by the external partner)
- The workflow for `in` and `out` letters is now the same. 
  - This way we can track letters sent until we get confrimation of reception and manually set it to 'received'
  - `snd_rec_date` is replaced by fields `snd_date` and `rec_date`
  - `snd_date` is the date the letter was sent either by us (`out`) or by the external sender (`in`)
  - `rec_date` is the date the letter was received by us (`in`) or the date the recipient confirmed (`out`)
- Simplified workflow
  - Removed `created` and `validated` states.
  - Current workflow is: 
    - `Draft -> Sent -> Received / Rec. Returned / Rec. Bad`
    - `Cancelled -> Back to Draft`
- Groups are added to show/hide some fields for extended usage
  - group to show/hide parent and childs ("thread")
  - group to show/hide reassignments
- Sender and Recipient are now defaulted to current company, depending on the letter move (`in` or `out`)
- It no longer add subscribers when changing the state. To my understanding, it was annoying.
- Some other changes..

**Comment by ivantodorovich** [comment]
I'm need to make the module to extend the workflow. It will re-implement the 'created' and 'validated' states.

I still don't completely understand the use case, though.. I mean, it didn't have security groups, so any user was able to create or validate any letter. If it was only to track the letter prep states before sending, the same could be achived by just using internal notes or messages.

Even so, it seems to be a very company-specific case. I think it's very possible no one will use it anymore (and ever). 

Should I do it the same? I don't mind.. I just don't like the idea of useless modules.

**Comment by ivantodorovich** [comment]
@max3903 I'd like to hear your thoughts too

**Comment by hbrunn** [comment]
what's the state of this PR? Can we review?

**Comment by hbrunn** [comment]
(asking because I'm in the process of adding a 'done' state and support for needaction, and it would make sense to get your refactoring merged first of course)

File: lettermgmt/migrations/8.0.2.0/pre-migration.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+def migrate(cr, version):
+    if version is None:
+        return
+
+    # Migrate 'type' to 'type_id'
+    cr.execute("""ALTER TABLE res_letter
+                  RENAME COLUMN type to type_id""")
```
Comment (hbrunn) [inline lettermgmt/migrations/8.0.2.0/pre-migration.py]: you should borrow openupgrade's code for renaming a model here: https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade.py#L328 to get links from other models (mostly) right after you renamed `letter.class`

File: lettermgmt/migrations/8.0.2.0/post-migration.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+def migrate(cr, version):
+    if version is None:
+        return
+
+    # Migrate states ['created', 'validated'] back to 'draft'
+    cr.execute("""UPDATE res_letter SET state = 'draft'
+                  WHERE state IN ('created', 'validated')""")
+
+    # Migrate 'letter.class' to 'letter.category'
+    cr.execute("""INSERT INTO letter_category SELECT * FROM letter_class""")
```
Comment (hbrunn) [inline lettermgmt/migrations/8.0.2.0/post-migration.py]: better to this in pre-migrate, see below

**Comment by ivantodorovich** [comment]
> @hbrunn 
> what's the state of this PR? Can we review?
> (asking because I'm in the process of adding a 'done' state and support for needaction, and it would make sense to get your refactoring merged first of course)

Well, I kinda stop working on it because of the low attention it got.
But I believe this PR was almost ready, except for the 'advanced workflow' support, which is removed and it should be implemented in a separate module because of its rare usage.. but maybe no one will ever use it.

I'll take a look at openupgrade's code, as you suggested, and make the changes later this week.

**Comment by hbrunn** [comment]
that's probably because of the work in progress label - I changed this to needs review to invite other reviewers

File: lettermgmt/models/res_letter.py [hunk 3]
```diff
@@ -1,165 +1,200 @@
...
-        """Put the state of the letter into Validated"""
-        for letter in self.browse(cr, uid, ids, context=context):
-            self.write(
-                cr, uid, [letter.id], {'state': 'validated'}, context=context)
+    @api.multi
+    def action_cancel_draft(self):
+        """ Go from cancelled state to draf state """
+        self.write({'state': 'draft'})
+        self.delete_workflow()
+        self.create_workflow()
```
Comment (hbrunn) [inline lettermgmt/models/res_letter.py]: this model doesn't define a workflow (in the sense of odoo's workflows), so why do you need this?

File: lettermgmt/models/res_letter.py [hunk 4]
```diff
@@ -1,165 +1,200 @@
...
-                },
-                context=context)
-            if letter.sender_partner_id:
-                letter.message_subscribe([letter.sender_partner_id.id])
+    @api.multi
+    def action_send(self):
+        """ Put the state of the letter into sent """
+        self.write({
+            'state': 'sent',
+            'snd_date': self.snd_date or fields.Date.today()
```
Comment (hbrunn) [inline lettermgmt/models/res_letter.py:169]: this is going to break for n > 1. You really need the loop

File: lettermgmt/models/res_letter.py [hunk 5]
```diff
@@ -1,165 +1,200 @@
...
-        """Put the state of the letter into Received but Returned"""
-        for letter in self.browse(cr, uid, ids, context=context):
-            self.write(
-                cr, uid, [letter.id], {'state': 'rec_ret'}, context=context)
+    @api.multi
+    def action_received(self):
+        """ Put the state of the letter into Received """
+        self.write({
+            'state': 'rec',
+            'rec_date': self.rec_date or fields.Date.today()
```
Comment (hbrunn) [inline lettermgmt/models/res_letter.py:178]: same here

File: lettermgmt/models/res_letter.py [hunk 6]
```diff
@@ -1,165 +1,200 @@
...
-        """Put the state of the letter into Received but Damaged"""
-        for letter in self.browse(cr, uid, ids, context=context):
-            self.write(
-                cr, uid, [letter.id], {'state': 'rec_bad'}, context=context)
+    @api.multi
+    def action_rec_ret(self):
+        """ Put the state of the letter into Received but Returned """
+        self.write({
+            'state': 'rec_ret',
+            'rec_date': self.rec_date or fields.Date.today()
```
Comment (hbrunn) [inline lettermgmt/models/res_letter.py:187]: and here

File: lettermgmt/models/res_letter.py [hunk 7]
```diff
@@ -1,165 +1,200 @@
...
-        """Put the state of the letter into draft"""
-        for letter in self.browse(cr, uid, ids, context=context):
-            self.write(
-                cr, uid, [letter.id], {'state': 'draft'}, context=context)
+    @api.multi
+    def action_rec_bad(self):
+        """ Put the state of the letter into Received but Damaged """
+        self.write({
+            'state': 'rec_bad',
+            'rec_date': self.rec_date or fields.Date.today()
```
Comment (hbrunn) [inline lettermgmt/models/res_letter.py:196]: same

File: lettermgmt/views/letter_category_view.xml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+     © 2009 Parthiv Patel, Tech Receptives
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<openerp>
+  <data>
+
+    <!-- Views -->
+
+    <record model="ir.ui.view" id="letter_category_tree_view">
+      <field name="name">letter.category.tree</field>
```
Comment (hbrunn) [inline lettermgmt/views/letter_category_view.xml]: this is nearly what odoo autogenerates as name when you leave out the field, so you could do just that

File: lettermgmt/models/res_letter.py (hunk 3 above)
Comment (ivantodorovich) [reply lettermgmt/models/res_letter.py]: At the time of this PR, I was just starting to learn odoo. I guess I've just copied this from another module's code, and it doesn't apply here at all.

File: lettermgmt/models/res_letter.py (hunk 4 above)
Comment (ivantodorovich) [reply lettermgmt/models/res_letter.py:169]: Yeah.. sorry about that. I should've reviewed past me myself.
I'll fix all of this later this week, probably Friday.

File: lettermgmt/migrations/8.0.2.0/pre-migration.py (hunk 1 above)
Comment (ivantodorovich) [reply lettermgmt/migrations/8.0.2.0/pre-migration.py]: Done

File: lettermgmt/migrations/8.0.2.0/post-migration.py (hunk 2 above)
Comment (ivantodorovich) [reply lettermgmt/migrations/8.0.2.0/post-migration.py]: I couldn't do it in pre-migrate. I tried a few approaches and it kept giving me errors.

**Comment by ivantodorovich** [comment]
I need help with migration. I'm not happy with how it is right now. I'm not sure if it works properly either.

**Comment by hbrunn** [comment]
@ivantodorovich just from looking at the code, the migration seems okay. What is it you're not happy about?

**Comment by ivantodorovich** [comment]
This right here:

`cr.execute("""INSERT INTO letter_category SELECT * FROM letter_class""")`

openupgrade renames the table instead of copying all the records from the old table into the new table.
I tried to do it that way, in pre-migration, but odoo won't accept it for some reason I couldn't figure out.

Anyway, considering this module's popularity, I guess no one will ever care about that, as long as it works.

**Comment by hbrunn** [comment]
what error message do you get here? And did you prepare your test database accordingly?

**Comment by rafaelbn** [comment]
Do you need any functional help here @ivantodorovich @hbrunn ?

**Comment by ivantodorovich** [comment]
I'm sorry for not responding earlier. 
I haven't been able to make up the time this week.

I'll test again the migration code on a clean database and see if I can improve it. 
Any help reviewing the migration would be great.

**Comment by pedrobaeza** [comment]
This PR is hard to review, but I can trust @hbrunn if he says this is OK

**Comment by pedrobaeza** [comment]
@ivantodorovich, can you squash a bit the commits?

@hbrunn, any comments?

**Comment by ivantodorovich** [comment]
Done. Squashed into a single commit.

**Comment by hbrunn** [comment]
my comment is let's merge as soon as @ivantodorovich has fixed the last travis nags: https://travis-ci.org/OCA/crm/jobs/168795594#L314

**Comment by hbrunn** [comment]
better link: https://travis-ci.org/OCA/crm/jobs/168795594#L307-L323

**Comment by ivantodorovich** [comment]
Done, sorry about that!

**Comment by pedrobaeza** [comment]
Thanks for your efforts!

**Pull request merged**